File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/stack/P496_NextGreaterElementI.java
package com.dsa.leetcode.labuladong.stack;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashMap;
import java.util.Map;

public class P496_NextGreaterElementI {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> map = new HashMap<>(nums2.length);
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = nums2.length - 1; i >= 0; i--) {
            while (!stack.isEmpty() && nums2[i] > stack.peek()) {
                stack.pop();
            }
            map.put(nums2[i], stack.isEmpty() ? -1 : stack.peek());
            stack.push(nums2[i]);
        }
        int[] res = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            res[i] = map.get(nums1[i]);
        }
        return res;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/stack/P496_NextGreaterElementI.py
# author: tobebetter9527
# since: 2024/4/28 18:02
from typing import List


class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        res = {}
        stack = []
        for num in reversed(nums2):
            while stack and num > stack[-1]:
                stack.pop()
            res[num] = -1 if not stack else stack[-1]
            stack.append(num)
        return [res[num] for num in nums1]
